src/azul/service/manifest_service.py

```diff
@@ -705,10 +707,75 @@ def write_to(self, output: IO[str]) -> Optional[str]:
             url = furl(config.service_endpoint(),
                        path=f'/repository/files/{uuid}',
                        args=dict(version=version, catalog=self.catalog))
+            name = self._sanitize_path(name)
             output.write(f'url={self._option(url.url)}\n'
                          f'output={self._option(name)}\n\n')
         return None
 
+    control_characters = re.compile(r'[\x00-\x1f\\]')
+
+    replaceable_characters = re.compile(r'[<>:"|?*]')
+
+    valid_regex_sequence = re.compile(r'''
+        (\.?[^./ ])([^/]*[^./ ])? # sequences with '/', ' ', '.' zero or more times can compromise system files & paths,
+        (/(\.?[^./ ])([^/]*[^./ ])?)* # they are only allowed in between characters, not at start or end of a sequence 
+    ''', re.X)
+
+    disallowed_path_components = {
+        'CON', 'PRN', 'AUX', 'NUL',
+        *(f'{cmd}{i}' for cmd in ['COM', 'LPT'] for i in range(1, 10))
+    }
+
+    @classmethod
+    def _sanitize_path(cls, path: str) -> str:
+        """
+        >>> f = CurlManifestGenerator._sanitize_path
+        >>> f('foo/bar/\\x1F/file') # doctest: +NORMALIZE_WHITESPACE
+        Traceback (most recent call last):
+        ...
+        azul.RequirementError: ('Invalid file name.', 'foo/bar/\\x1f/file',  'Control characters and backslash are not
+        allowed.', '\\x1f')
+
+        >>> f('foo/bar/\\/file') # doctest: +NORMALIZE_WHITESPACE
+        Traceback (most recent call last):
+        ...
+        azul.RequirementError: ('Invalid file name.', 'foo/bar/\\\\/file', 'Control characters and backslash are not
+        allowed.', '\\\\')
+
+        >>> f('foo/bar/COM6/file')
+        Traceback (most recent call last):
+        ...
+        azul.RequirementError: ('Path component not allowed', {'COM6'})
+
+        >>> f('foo/bar/ / baz/file')
+        Traceback (most recent call last):
+        ...
+        azul.RequirementError: ('Name sequence not allowed in file name', 'foo/bar/ / baz/file')
+
+        >>> f('foo/bar?/fi*le|')
+        'foo/bar_/fi_le_'
+
+        >>> f('foo/bar/file.fastqgz')
+        'foo/bar/file.fastqgz'
+        """
+        match = cls.control_characters.search(path)
+        require(match is None,
+                'Invalid file name.',
+                path,
+                'Control characters and backslash are not allowed.',
+                path[match.start()]) if match else None
```

Comment:
I wasn't to sure how to go about this conditional, `match.start()` is only valid when `match` is of `Match` type, and that only happens if we have a "disallowed sequence" do to it containing control characters. In any other cases, `match` is `NoneType` which is why I had to make the `require(…` conditional.

Reply:
You don't *have* to use require/reject.